crates/pico_macros/src/memo.rs

```diff
@@ -124,27 +132,90 @@ pub(crate) fn memo(args: TokenStream, item: TokenStream) -> TokenStream {
+                        quote! { param_id.into() }
+                    };
+                    quote! {
+                        let #arg: #ty = {
+                            let param_id = derived_node_id.params[#i];
+                            #maybe_ref
+                        };
+                    }
+                }
+                ArgType::MemoRef => {
+                    let maybe_ref = if matches!(**ty, syn::Type::Reference(_)) {
+                        quote! { &::pico::MemoRef::new(#db_arg, param_id.into()) }
+                    } else {
+                        quote! { ::pico::MemoRef::new(#db_arg, param_id.into()) }
+                    };
+                    quote! {
+                        let #arg: #ty = {
+                            let param_id = derived_node_id.params[#i];
+                            #maybe_ref
+                        };
+                    }
+                }
+                ArgType::Other => {
+                    let target_type = if let syn::Type::Reference(ref reference) = **ty {
+                        &reference.elem
+                    } else {
+                        ty
+                    };
+                    let binding_expr = match **ty {
+                        syn::Type::Reference(_) => quote!(inner),
+                        _ => quote!(inner.clone()),
+                    };
+                    quote! {
+                        let #arg = {
+                            let param_ref = ::pico::macro_fns::get_param(#db_arg, derived_node_id.params[#i])
+                                .expect("param should exist. This is indicative of a bug in Pico.");
+                            let inner = param_ref.downcast_ref::<#target_type>()
+                                .expect("param type must be correct. This is indicative of a bug in Pico.");
+                            #binding_expr
+                        };
+                    }
+                }
+            }
+        });
+
     let output = quote! {
         #(#attrs)*
         #vis #new_sig {
-            let param_id = ::pico::macro_fns::intern_param(#db_arg, (#(#other_args.clone(),)*));
-            let derived_node_id = ::pico::DerivedNodeId::new(#fn_hash.into(), param_id);
-            ::pico::memo(#db_arg, derived_node_id, ::pico::InnerFn::new(|#db_arg, param_id| {
-                let param_ref = ::pico::macro_fns::get_param(#db_arg, param_id)
-                    .expect("param should exist. This is indicative of a bug in Pico.");
-                let (#(#unpacked_args,)*) = {
-                    let (#(#inner_args,)*) = param_ref
-                        .downcast_ref::<(#(#get_ref_types,)*)>()
-                        .expect("param type must to be correct. This is indicative of a bug in Pico.");
-                    (
-                        #(#reconstructed_args,)*
-                    )
-                };
+            let mut param_ids = ::pico::macro_fns::init_param_vec();
```

Comment:
nit: this can probably just be inlined here, right?

Reply:
If we ok to reexport `ArrayVec`

Reply:
Oh I see, then it makes sense to use `init_param_vec`.